eras/conway/impl/src/Cardano/Ledger/Conway/Rules/Ledger.hs

```diff
@@ -396,16 +396,26 @@ ledgerTransition = do
               , certState
               , StrictSeq.fromStrict $ txBody ^. certsTxBodyL
               )
-        let wdrlAddrs = Map.keysSet . unWithdrawals $ tx ^. bodyTxL . withdrawalsTxBodyL
-            wdrlCreds = Set.map raCredential wdrlAddrs
-            dUnified = dsUnified $ certDState certStateAfterCERTS
-            delegatedAddrs = DRepUView dUnified
-
-        -- TODO: Finish this implementation once we are in bootstrap phase:
-        -- https://github.com/IntersectMBO/cardano-ledger/issues/4092
-        when False $ do
-          all (`UMap.member` delegatedAddrs) wdrlCreds
-            ?! ConwayWdrlNotDelegatedToDRep (wdrlCreds Set.\\ Map.keysSet (dRepMap dUnified))
+
+        let checkWdrlsKeyHashesDelegated = do
+              let
+                isKeyHash = \case
+                  KeyHashObj _ -> True
+                  _ -> False
+                wdrlsKeyHashes =
+                  Set.filter isKeyHash $
+                    Set.map raCredential $
+                      Map.keysSet $
+                        unWithdrawals $
+                          tx ^. bodyTxL . withdrawalsTxBodyL
+                dUnified = dsUnified $ certDState certStateAfterCERTS
+                notDelegated = wdrlsKeyHashes Set.\\ Map.keysSet (dRepMap dUnified)
```

Comment:
Avoid redundant conversions from a Map to a Set, especially for unified map, it is huge:
```suggestion
                        unWithdrawals $
                          tx ^. bodyTxL . withdrawalsTxBodyL
                dUnified = dsUnified $ certDState certStateAfterCERTS
                notDelegated = Map.keysSet (wdrlsKeyHashes Map.\\ dRepMap dUnified)
```

Reply:
To define `notDelegated` as above, I defined 
`wdlrsKeyHashes` like this: 
```
          wdrlsKeyHashes =
                Map.mapKeys raCredential $
                  Map.mapMaybeWithKey
                    (\k v -> v <$ credKeyHash (raCredential k))
                    wdrls
```
Is there a better way to transform a Map that combines: filtering out `Nothing` values with mapping the keys?   Because the keys have to be `Credential`  in order to do `\\ drepMap unified` 

I could use `foldlWithKey` instead,  like this, to traverse the map just once :

```     
              wdrlsKeyHashes = Map.foldlWithKey' accum Map.empty
                                  (unWithdrawals $ tx ^. bodyTxL . withdrawalsTxBodyL)
              accum res k v = case credKeyHash (raCredential k) of
                Just _ -> Map.insert (raCredential k) v res
                Nothing -> res
```